A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+Prior to this proposal, gRPC supported two values for this field,
+`ROUND_ROBIN` and `RING_HASH`.  With this proposal, we will add support for an
+additional policy, `LEAST_REQUEST`.
+
+The configuration for the Least Request LB policy is the
+[`least_request_lb_config`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L916).
+The field is optional; if not present, defaults will be assumed for all
+of its values.  gRPC will support the
+[`choice_count`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L346)
+field the same way that Envoy does with the exception of allowed values
+(More details further down in [LB Policy Config](#lb-policy-config)).
+The
+[`active_request_bias`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L371)
+field will be ignored as it is only used in the "*all weights not equal*" (weighted least request)
+case which will not be supported as part of this proposal.
+
+#### Changes to the `xds_cluster_resolver` Policy
+
+As a result of
+[gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md),
+the `xds_cluster_resolver` policy implementation will need additional logic to understand the setup for `LEAST_REQUEST`
+in addition to `ROUND_ROBIN` and `RING_HASH`.
+For `LEAST_REQUEST`, the policy name will be `"LEAST_REQUEST"`, and the config will be the one for the `least_request_experimental` LB Policy described below.
+When the `xds_cluster_resolver` policy sees `"LEAST_REQUEST"`, it will also assume that the
+locality-picking policy is `weighted_target` and the endpoint-picking policy is `least_request_experimental`.
+
+#### `least_request_experimental` LB Policy
+
+We will implement a `least_request_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
+for each pick by the picker.
+Envoy currently supports any `choice_count` value greater than two.
+In the `least_request_experimental` policy however, only an effective `choice_count` in the
+range `[2, 10]` will be supported. If a `LeastRequestLoadBalancingConfig` with a `choice_count > 10`
+is received, the `least_request_experimental` policy will set `choice_count = 10`.
+
+##### Subchannel state handling
```

Comment:
Hmm, it was tricky to find the equivalent logic for the different languages and also it's hard for me to tell if there are any slight discrepancies between them.
The [java implementation](https://github.com/grpc/grpc-java/blob/fe9026ed8ae5025e92d452e077bfa0b329c123a0/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java#L138-L156) has been the easiest to follow for me so if I just where to explain the "Subchannel state handling" there it would be something like:

```
1. New subchannels begin with the `IDLE` state and are immediately requested to connect.
2. When a subchannel enters either `TRANSIENT_FAILURE` or `IDLE` states, a re-resolve is triggered.
3. When a subchannel enters the `IDLE` state, it's immediately requested to connect.
4. When a subchannel enters the `TRANSIENT_FAILURE` state, it can only transition directly to the `READY` state.
5. Removed subchannels cause a disconnect and release of the underlying resources.
```

Possibly it's better to just describe it like this here instead of trying to find/link to implementations in each language.

Reply:
Linking to the full files is probably fine.  Honestly, I don't really care about linking to implementations at all, so you could just remove that if you want, although I don't object to the links being here.  But the important thing is to describe in text the required behavior of the `least_request` policy, independently of any existing implementations.